src/swarm-engine/Swarm/Game/Step/Flood.hs

```diff
@@ -0,0 +1,110 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Implementation of the 'Swarm.Language.Syntax.Volume' command for robots.
+--
+-- Note: If the robot is currently on an unwalkable cell (which may happen in
+-- the case of teleportation or if an entity is placed or pushed into its cell),
+-- the volume shall be zero.
+module Swarm.Game.Step.Flood (
+  floodFill,
+) where
+
+import Control.Effect.Lens
+import Control.Lens (makeLenses, (%~), (&))
+import Data.HashSet (HashSet)
+import Data.HashSet qualified as HashSet
+import Swarm.Game.Location
+import Swarm.Game.Step.RobotStepState
+import Swarm.Game.Step.Util (checkMoveFailureUnprivileged)
+import Swarm.Game.Step.Util.Inspect (getNeighborLocs)
+import Swarm.Game.Universe
+
+data FloodParms = FloodParms
+  { theSubworld :: SubworldName
+  , maxVisits :: Int
+  }
+
+data Tracking = Tracking
+  { visited :: HashSet Location
+  , floodPartition :: FloodPartition
+  }
+
+-- | We annotate each visited cell as
+-- being part of the boundary or the interior.
+-- This lets us:
+--
+-- 1. Use the interior cell count as a termination condition
+-- 2. Handle (eventual) cache invalidation differently for boundary
+--    members than interior members.
+data FloodPartition = FloodPartition
+  { _boundary :: HashSet Location
+  , _interior :: HashSet Location
+  }
+
+makeLenses ''FloodPartition
+
+-- |
+-- == Algorithm
+--
+-- Explore via DFS using a list as a stack.
+-- Each iteration examines a single cell.
+--
+-- 1. Mark the popped cell as visited, regardless of walkability.
+-- 2. Check popped cell for walkability
+-- 3. Add all neighbors that aren't already visited, regardless of walkability, to the stack.
+--    But unwalkable cells shall not produce neighbors and shall be marked with a boundary/interior distinction.
+floodRecursive ::
+  HasRobotStepState sig m =>
+  Tracking ->
+  [Location] ->
+  FloodParms ->
+  m (Maybe Int)
+floodRecursive tracking pending params =
+  case pending of
+    nextLoc : otherLocs ->
+      if interiorCount > maxVisits params
+        then return Nothing
+        else checkNeighbors nextLoc otherLocs
+    [] -> return $ Just interiorCount
+ where
+  interiorCount = HashSet.size $ _interior $ floodPartition tracking
+  checkNeighbors nextLoc otherLocs = do
+    isWalkable <- null <$> checkMoveFailureUnprivileged cosmicLoc
+    let candidateNeighbors =
+          if isWalkable
+            then map (view planar) $ getNeighborLocs cosmicLoc
+            else []
+        visitableNeighbors = filter (not . (`HashSet.member` visited tracking)) candidateNeighbors
+
+        -- It's cheaper to prepend the "visitableNeighbors" list because
+        -- it should in general be a shorter list than the "pending" list.
```

Comment:
This comment seems a little misleading --- the reason for prepending the `visitableNeighbors` is because we're treating `pending` as a stack rather than a queue, not because it is more efficient (though I agree it will be).  Regardless of which list is shorter, prepending the `visitableNeighbors` will take constant time per new neighbor, and hence $O(n)$ overall since each vertex occurs as a new neighbor at most once. Whereas appending `visitableNeighbors` will take time linear in the number of vertices in the queue, but the same vertices might be in the queue for multiple iterations, so we can end up doing more than $O(n)$ work appending.

Reply:
> the reason for prepending the `visitableNeighbors` is because we're treating `pending` as a stack rather than a queue, not because it is more efficient (though I agree it will be)

Well, I originally went about this intending to implement BFS, until I got to this line and realized that **appending** to the `pending` list would be expensive (and I couldn't be bothered to use `Seq` instead of `[]`).  So I decided to **prepend** instead, which turned it into DFS.